pulsar-client/src/main/java/org/apache/pulsar/client/impl/PartitionedConsumerImpl.java

```diff
@@ -418,23 +449,23 @@ private ConsumerConfiguration getInternalConsumerConfig() {
 
     @Override
     public void redeliverUnacknowledgedMessages() {
-        for (ConsumerImpl c : consumers) {
-            c.redeliverUnacknowledgedMessages();
+        synchronized (this) {
+            incomingMessages.clear();
+            unAckedMessageTracker.clear();
+            for (ConsumerImpl c : consumers) {
+                c.redeliverUnacknowledgedMessages();
+            }
         }
+
     }
 
     @Override
     public void redeliverUnacknowledgedMessages(Set<MessageIdImpl> messageIds) {
+        removeExpiredMessagesFromQueue(messageIds);
+        Map<Integer, Set<MessageIdImpl>> setMap = messageIds.stream()
+                .collect(Collectors.groupingBy(MessageIdImpl::getPartitionIndex, Collectors.toSet()));
         for (ConsumerImpl c : consumers) {
-            Set<MessageIdImpl> consumerMessageIds = new HashSet<>();
-            messageIds.removeIf(messageId -> {
-                if (messageId.getPartitionIndex() == c.getPartitionIndex()) {
-                    consumerMessageIds.add(messageId);
-                    return true;
-                }
-                return false;
-            });
-            c.redeliverUnacknowledgedMessages(consumerMessageIds);
+            c.redeliverUnacknowledgedMessages(setMap.get(c.getPartitionIndex()));
```

Comment:
I believe that these code could lead to a bug.

setMap might not have expired messages for all partitions, therefore `setMap.get(c.getPartitionIndex()));` is going to be `null` generating some possible `NullPointerException` on `redeliverUnacknowledgedMessages` method.
I imagine two possible solutions:
1. instead of using `.get` use `.getOrDefault` checking if default was obtained before calling `redeliverUnacknowledgedMessages` method.
2. Just check if setMap has proper partition index.
```java
int cPartitionIdx = c.getPartitionIndex();
if (setMap.containsKey(cPartitionIdx)) {
    c.redeliverUnacknowledgedMessages(setMap.get(cPartitionIdx));
}
```

Reply:
nice catch, fixed